### PR TITLE
Feature/exception factory

### PR DESCRIPTION
**Pull request by OGKevin**
#49 

- [x] Exception factory review

- [x] Explain exceptions in a .md file

- [x] Final review

### CODE REVIEW

File: src/Exception/ApiException.php [hunk 1]
```diff
@@ -3,6 +3,35 @@
 
 /**
  */
-class ApiException extends BunqException
+class ApiException extends \Exception
```
Comment (dnl-blkv) [inline src/Exception/ApiException.php]: Let us please do `use Exception;` and then `Exception` with no preceding `\` :)

File: src/Exception/ExceptionFactory.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * The first item index in an array.
+     */
+    const INDEX_FIRST = 0;
+
+    /**
+     * @param string $message
+     * @param array $args
+     *
+     * @return BadRequestException|ForbiddenException|MethodNotAllowedException|NotFoundException|PleaseContactBunqException|ToManyRequestsException|UnauthorizedException|UnknownApiErrorException
```
Comment (dnl-blkv) [inline src/Exception/ExceptionFactory.php]: Just `ApiException` would be good :P

File: src/Exception/ToManyRequestsException.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+namespace bunq\Exception;
+
+/**
+ */
+class ToManyRequestsException extends ApiException
```
Comment (dnl-blkv) [inline src/Exception/ToManyRequestsException.php]: One `o` refuged it seems.

File: src/Exception/ApiException.php [hunk 4]
```diff
@@ -3,6 +3,35 @@
...
+     * @var int
+     */
+    private $responseCode;
+
+    /**
+     * ApiException constructor.
+     *
+     * @param int $responseCode
+     */
+    public function __construct($message, $args = [])
```
Comment (dnl-blkv) [inline src/Exception/ApiException.php]: Please add strict types for arguments (PHP 7.0 style) and doc block params. Also, why are we using `$args`? Is it necessary here? If so, please add a comment explaining this to the constructor header.

File: src/Exception/ApiException.php [hunk 5]
```diff
@@ -3,6 +3,35 @@
...
+    {
+        $this->responseCode = $args[self::INDEX_FIRST];
+
+        parent::__construct(vsprintf($message, $args));
+    }
+
+    /**
+     * @return int
+     */
+    public function getResponseCode()
```
Comment (dnl-blkv) [inline src/Exception/ApiException.php]: Please add PHP 7.0 strict response type (`: int`)

File: src/Exception/ExceptionFactory.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public static function createExceptionForResponse($message, $args = [])
```
Comment (dnl-blkv) [inline src/Exception/ExceptionFactory.php]: Same question: do we need args here?

Also please add strict param types and the response type :)

File: src/Http/Handler/ResponseHandlerError.php [hunk 7]
```diff
@@ -60,15 +61,15 @@ public function execute(ResponseInterface $response)
                     $errorDescriptions[] = $error[self::FIELD_ERROR_DESCRIPTION];
                 }
 
-                throw new ApiException(
+                throw ExceptionFactory::createExceptionForResponse(
                     self::ERROR_FROM_JSON,
```
Comment (dnl-blkv) [inline src/Http/Handler/ResponseHandlerError.php]: I know it is not yours, but please make the interface of `createExceptionForResponse` consistent with that in all the other SDKs if possible :)

File: src/Http/Handler/ResponseHandlerError.php [hunk 8]
```diff
@@ -60,15 +61,15 @@ public function execute(ResponseInterface $response)
...
                     [
                         $response->getStatusCode(),
                         implode(self::SEPERATOR_ERROR, $errorDescriptions),
                     ]
                 );
             } else {
-                throw new ApiException(
+                throw ExceptionFactory::createExceptionForResponse(
                     self::ERROR_UNKNOWN,
```
Comment (dnl-blkv) [inline src/Http/Handler/ResponseHandlerError.php]: Also, please make error handling here consistent with what we have in e.g. Python :)

File: src/Exception/ToManyRequestsException.php (hunk 3 above)
Comment (OGKevin) [reply src/Exception/ToManyRequestsException.php]: OMG this is in all the SDK's 😁 🙈

File: src/Exception/ApiException.php [hunk 9]
```diff
@@ -1,8 +1,34 @@
...
-class ApiException extends BunqException
+class ApiException extends Exception
 {
+    /**
+     * @var int
+     */
+    private $responseCode;
+
+    /**
+     * ApiException constructor.
```
Comment (dnl-blkv) [inline src/Exception/ApiException.php]: This comment can go :)

File: src/Exception/ApiException.php [hunk 10]
```diff
@@ -1,8 +1,34 @@
...
+     *
+     * @param int $responseCode
+     */
+    public function __construct($message, $responseCode)
```
Comment (dnl-blkv) [inline src/Exception/ApiException.php]: Please add strict types and docs for both parameters!

File: src/Exception/BunqException.php [hunk 11]
```diff
@@ -7,10 +7,10 @@ class BunqException extends \Exception
 {
     /**
      * @param string $message
-     * @param string[] $args
+     * @param string[] $responseCode
      */
-    public function __construct($message, $args = [])
+    public function __construct($message, $responseCode = [])
```
Comment (dnl-blkv) [inline src/Exception/BunqException.php]: Please add strict types for both parameters; Also `$responseCode` shouldn't be equal to an array I believe :D

File: src/Exception/ExceptionFactory.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    const FORMAT_RESPONSE_CODE_LINE = 'HTTP Response Code: %s';
+    const GLUE_ERROR_MESSAGES = "\n";
+
+    /**
+     * The first item index in an array.
+     */
+    const INDEX_FIRST = 0;
+
+    /**
+     * @param array $messages
```
Comment (dnl-blkv) [inline src/Exception/ExceptionFactory.php]: Please put the strict type here :)

File: src/Exception/ExceptionFactory.php [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+     * @param int $responseCode
+     *
+     * @return ApiException
+     */
+    public static function createExceptionForResponse($messages, $responseCode)
```
Comment (dnl-blkv) [inline src/Exception/ExceptionFactory.php]: tyyypes (both params and return ones)

File: src/Http/ApiClient.php [hunk 14]
```diff
@@ -188,7 +188,7 @@ private function initializeHttpClient()
                     self::OPTION_HANDLER => $middleware,
                     self::OPTION_VERIFY => true,
                     self::OPTION_CURL => [
-                        CURLOPT_PINNEDPUBLICKEY => $this->determinePinnedServerPublicKey(),
+                        //CURLOPT_PINNEDPUBLICKEY => $this->determinePinnedServerPublicKey(),
```
Comment (dnl-blkv) [inline src/Http/ApiClient.php]: Wut? Why is it commented out? :D

Comment (OGKevin) [reply src/Http/ApiClient.php]: Shit yea haha, m y curl doesn’t work neither and didn’t fix it 😁. Therefore I must comment this line out for the sdk to work. Obviously forgot to revert it.

File: src/Exception/ApiException.php (hunk 10 above)
Comment (OGKevin) [reply src/Exception/ApiException.php]: Crap, I’m really getting old indeed xD

**Comment by OGKevin** [comment]
@dnl-blkv please 👀

File: src/Exception/EXCEPTION.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+     * @var int
+     */
+    private $responseCode;
+
+    /**
+     * @param string $message
+     * @param int $responseCode
+     */
+    public function __construct(string $message, int $responseCode)
+    {}
```
Comment (dnl-blkv) [inline src/Exception/EXCEPTION.md]: Please make it clear that some code is hidden here :)

**Comment by OGKevin** [comment]
@dnl-blkv all yours 🎖

File: src/Exception/EXCEPTION.md [hunk 16]
```diff
@@ -0,0 +1,84 @@
+## Exceptions
+When you make a request via the SDK, there is a chance of request failing
+due to various reasons. When such a failure happens, an exception
+corresponding to the error occurred is raised.
```
Comment (dnl-blkv) [inline src/Exception/EXCEPTION.md]: Exceptions are thrown in PHP, not raised! :)

File: src/Exception/EXCEPTION.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+
+----
```
Comment (dnl-blkv) [inline src/Exception/EXCEPTION.md]: Do we need these `----` ?

File: src/Exception/EXCEPTION.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+     * @return int
+     */
+    public function getResponseCode(): int
+    {
+        return $this->responseCode;
+    }
+}
+```
+The `Exception` class which is being extended has an `getMessage()` method which is `final` and therefore cannot be
+overwritten.
```
Comment (dnl-blkv) [inline src/Exception/EXCEPTION.md]: "overridden" please :P

File: src/Exception/EXCEPTION.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+
+const API_KEY = 'Some invalid API key '; 
+const DEVICE_DESCRIPTION = 'This will cause BadRequestException to be thrown.';
+
+try{
+    //Make a call that might fail
+    $apiContext = ApiContext::create(BunqEnumApiEnvironmentType::SANDBOX(), API_KEY, DEVICE_DESCRIPTION);
+} catch (BadRequestException $error){
+    //Do something if exception is thrown.
```
Comment (dnl-blkv) [inline src/Exception/EXCEPTION.md]: It is a minor thing, but here and in all the inline comments (`// Comment`) please make sure there is space between `//` and the first letter!

File: tests/Model/Generated/Endpoint/SessionTest.php [hunk 20]
```diff
@@ -31,6 +36,6 @@ public function testDeleteSession()
      */
     public static function tearDownAfterClass()
     {
-        static::ensureApiContextValid();
+        unlink(static::FILENAME_CONTEXT_CONFIG);
```
Comment (dnl-blkv) [inline tests/Model/Generated/Endpoint/SessionTest.php]: Oops! I must have missed it at some earlier point. Let us please use `self::` for constants :)

File: src/Exception/EXCEPTION.md (hunk 17 above)
Comment (OGKevin) [reply src/Exception/EXCEPTION.md]: This makes a it more beautiful :P Its like a paragraph separator. View it as an actual markdown and you shall see 😁

**Comment by dnl-blkv** [comment]
@andrederoos

**Comment by DennisSnijder** [comment]
@dnl-blkv @andrederoos new release? 😄

**Comment by dnl-blkv** [comment]
@DennisSnijder yep, planned for tomorrow :)

**Comment by DennisSnijder** [comment]
@dnl-blkv Awesome! thanks 👍

File: src/Http/Handler/ResponseHandlerError.php [hunk 21]
```diff
@@ -45,37 +29,47 @@ class ResponseHandlerError extends ResponseHandlerBase
...
+     *
+     * @return string[]
+     */
+    private function fetchErrorMessages(ResponseInterface $response): array
+    {
+        $responseBody = $response->getBody();
+        $responseBodyInJson = json_decode($responseBody, true);
 
-                $errorDescriptions = [];
+        if ($responseBodyInJson != false){
```
Comment (andrederoos) [inline src/Http/Handler/ResponseHandlerError.php]: @OGKevin `!==`. lets not fall in the php-weak-typing-trap

Comment (OGKevin) [reply src/Http/Handler/ResponseHandlerError.php]: Ahh 🙈 nice spot !

**Pull request merged**